### ISSUE TITLE
Typscript error: Type 'typeof IndexedDBProvider' is not assignable to type 'typeof StorageProvider'

### ISSUE THREAD

**Issue by mr-zwets** [mainnet-cash/mainnet-js#199]
error message VS code:
> type 'typeof IndexedDBProvider' is not assignable to type 'typeof StorageProvider'.
  Types of construct signatures are incompatible.
    Type 'new (dbName: string) => IndexedDBProvider' is not assignable to type 'abstract new () => StorageProvider'.
      Target signature provides too few arguments. Expected 1 or more, but got 0.ts(2322)
(property) BaseWallet.StorageProvider?: typeof StorageProvider | undefined

code:
```ts
  // set mainnet-js config
  Config.EnforceCashTokenReceiptAddresses = true;
  BaseWallet.StorageProvider = IndexedDBProvider;
```

screenshot:
![image](https://github.com/mainnet-cash/mainnet-js/assets/53938059/9c85ec83-1899-4c11-8b52-d547d70d950d)

**Comment by Sydwell** [comment]
Yes, the constructor of IndexedDBProvider is definitely an issue.
As a work around I created my own FixedProvider.ts file.
`import Dexie from "dexie";
import { StorageProvider, WalletDbEntryI } from "mainnet-js";

export class FixedProvider extends Dexie implements StorageProvider {
    public db: Dexie.Table<WalletDbEntryI, number>;

  public constructor() {
    super('FixedName');
    this.version(3).stores({
      wallet: "name",
    });
    this.db = this.table("wallet");
  }

  public async init() {
    return this;
  }

  public async close() {
    return this;
  }

  public getInfo() {
    return "indexedDB";
  }

  public async addWallet(name: string, walletId: string): Promise<boolean> {
    return this.transaction("rw", this.db, async () => {
      if ((await this.db.where({ name: name }).count()) === 0) {
        await this.db.add({ name: name, wallet: walletId }).catch((e) => {
          throw Error(e);
        });
        return true;
      } else {
        return false;
      }
    }).catch((e) => {
      throw e.stack || e;
    });
  }

  public async getWallet(name: string): Promise<WalletDbEntryI | undefined> {
    let obj = await this.db.get({ name: name });
    if (obj) {
      return obj;
    } else {
      return;
    }
  }

  public async getWallets(): Promise<Array<WalletDbEntryI>> {
    let walletObjects = await this.transaction("r", this.db, async () => {
      return await this.db.where("id").above(0).toArray();
    });
    if (walletObjects) {
      const WalletArray: WalletDbEntryI[] = await Promise.all(
        walletObjects.map(async (obj: WalletDbEntryI) => {
          return obj;
        })
      );
      return WalletArray;
    } else {
      return [];
    }
  }

  public async updateWallet(name: string, walletId: string): Promise<void> {
    this.transaction("rw", this.db, async () => {
      const collection = this.db.where({ name: name });
      if ((await collection.count()) === 0) {
        return false;
      } else {
        const wallet = (await collection.first())!;
        await this.db
          .put({ id: wallet.id!, name: name, wallet: walletId }, wallet.id!)
          .catch((e) => {
            throw Error(e);
          });
        return true;
      }
    }).catch((e) => {
      throw e.stack || e;
    });
  }

  public async walletExists(name: string): Promise<boolean> {
    return (await this.getWallet(name)) !== undefined;
  }          

}`

**Comment by mainnet-pat** [comment]
I can not trigger this typecheck, but this diff should fix the issue @2qx 

```diff
diff --git a/packages/mainnet-js/src/db/StorageProvider.ts b/packages/mainnet-js/src/db/StorageProvider.ts
index 17074cf8..a194b379 100644
--- a/packages/mainnet-js/src/db/StorageProvider.ts
+++ b/packages/mainnet-js/src/db/StorageProvider.ts
@@ -1,6 +1,10 @@
 import { WalletDbEntryI } from "./interface.js";
 
 export default abstract class StorageProvider {
+  constructor(...args: any[]) {
+    throw new Error("StorageProvider is abstract and cannot be instantiated");
+  }
+
   /**
    * Ensure the database exists and is open
    * @throws {Error} If the wallet could not be opened.
```

**Comment by mainnet-pat** [comment]
Is it possible to make our tsconfig more strict with regards to these typechecks so we can catch them in our tests?

**Comment by mr-zwets** [comment]
This issue is fixed with the mainnet-js@^2.3.9 release, so I'll close this issue! 🎉
